src/NHibernate/AdoNet/GenericBatchingBatcher.cs

```diff
@@ -207,9 +207,10 @@ public int ExecuteNonQuery()
 				{
 					return 0;
 				}
+
 				var batcherCommand = _batcher.Driver.GenerateCommand(
 					_commandType,
-					_sql.ToSqlString(),
+					_batcher.GetSQL(_sql.ToSqlString()),
```

Comment:
But this is different behavior from other batchers. In other batchers interceptor is executed per each generated command in batch. And this one is executed once for full batch. This difference might be sensible...

Reply:
My understanding is that `IInterceptor.OnPrepareStatement` should be called before calling `DbCommand.Prepare`, that is why I added it here. Other batchers like `MySqlClientBatchingBatcher` or `SqlClientBatchingBatcher` are using a command set class (`XXXClientSqlCommandSet`) where each command is prepared separately and added to the set. `GenericBatchingBatcher` does not prepare each command separately, instead `DbCommand.Prepare` is called for the whole batch (only one `DbCommand` is used), so it already differs from other batchers. Calling `IInterceptor.OnPrepareStatement` for a statement that will not be prepared doesn't seem right to me.

Reply:
`IInterceptor.OnPrepareStatement` can contain some parsing logic and that logic might not expect that it's executed on batch. So this looks like breaking change to me...
That's my main concern.
Also I prefer old way as it allows to safely enable/disable batching and make sure your interceptor will work the same...

Reply:
> IInterceptor.OnPrepareStatement should be called before calling DbCommand.Prepare... Calling IInterceptor.OnPrepareStatement for a statement that will not be prepared doesn't seem right to me.

I don't get it. To me call to `DbCommand.Prepare` is some internal implementation detail. What importance of it? Comment for `OnPrepareStatement` says `Called when sql string is being prepared.`  - nothing about command preparation... Why do you think `OnPrepareStatement` is bound to be called before `DbCommand.Prepare`?

Reply:
> Why do you think OnPrepareStatement is bound to be called before DbCommand.Prepare?

By the name and how it is currently used in NHibernate. Prepared statement IMO is a quite known database feature, which can be also seen when searching it on google. The main reason why I called it only before calling `DbCommand.Prepare` is because I was trying to imitate other batchers, which are calling `OnPrepareStatement` for every generated command prior calling `DbCommand.Prepare` (called only when enabled). Do note that also `BasicResultSetsCommand` calls `OnPrepareStatement` for the whole batch of queries (see [InterceptorOnPrepareStatementTest](https://github.com/nhibernate/nhibernate-core/pull/2285/files#diff-cc4c115a1dff727c1516b41e876dfb4aR145)) and not for each query. 

> What importance of it?

Not much, as `OnPrepareStatement` is IMO mostly used for sql statement modifications and not for some kind of statistics, I updated the PR to call `OnPrepareStatement` for each statement separately.